File: app/src/main/java/com/example/administrator/essim/fragments/FragmentRank.java
package com.example.administrator.essim.fragments;

import android.content.Intent;
import android.content.SharedPreferences;
import android.graphics.Rect;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.design.widget.Snackbar;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.GridLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.Toolbar;
import android.view.Gravity;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.ProgressBar;

import com.example.administrator.essim.R;
import com.example.administrator.essim.activities.MainActivity;
import com.example.administrator.essim.activities.SearchActivity;
import com.example.administrator.essim.activities.SettingsActivity;
import com.example.administrator.essim.activities.ViewPagerActivity;
import com.example.administrator.essim.adapters.PixivAdapterGrid;
import com.example.administrator.essim.interf.OnItemClickListener;
import com.example.administrator.essim.network.AppApiPixivService;
import com.example.administrator.essim.network.RestClient;
import com.example.administrator.essim.response.IllustRankingResponse;
import com.example.administrator.essim.response.IllustfollowResponse;
import com.example.administrator.essim.response.IllustsBean;
import com.example.administrator.essim.response.RecommendResponse;
import com.example.administrator.essim.response.Reference;
import com.example.administrator.essim.utils.Common;
import com.nightonke.boommenu.BoomButtons.OnBMClickListener;
import com.nightonke.boommenu.BoomButtons.TextInsideCircleButton;
import com.nightonke.boommenu.BoomMenuButton;
import com.nightonke.boommenu.Util;

import java.util.List;
import java.util.Objects;

import retrofit2.Call;
import retrofit2.Callback;


/**
 * Created by Administrator on 2018/1/15 0015.
 */

public class FragmentRank extends BaseFragment {

    private static final String[] arrayOfRankMode = {"动态", "日榜", "周榜", "月榜", "新人", "原创", "男性向", "女性向"};
    private static final String[] modelist = {"day", "week", "month", "week_rookie", "week_original", "day_male", "day_female"};
    public int currentDataType = -1;
    private String next_url;
    private SharedPreferences mSharedPreferences;
    private PixivAdapterGrid mPixivAdapter;
    private RecyclerView mRecyclerView;
    private BoomMenuButton bmb;
    private ProgressBar mProgressBar;
    private Toolbar toolbar;
    private OnBMClickListener clickListener = index -> {
        switch (index) {
            case 0:
                if (currentDataType != -1) {
                    currentDataType = -1;
                    getFollowUserNewIllust();
                }
                break;
            case 1:
                if (currentDataType != 0) {
                    currentDataType = 0;
                    getRankList(currentDataType);
                }
                break;
            case 2:
                if (currentDataType != 1) {
                    currentDataType = 1;
                    getRankList(currentDataType);
                }
                break;
            case 3:
                if (currentDataType != 2) {
                    currentDataType = 2;
                    getRankList(currentDataType);
                }
                break;
            case 4:
                if (currentDataType != 3) {
                    currentDataType = 3;
                    getRankList(currentDataType);
                }
                break;
            case 5:
                if (currentDataType != 4) {
                    currentDataType = 4;
                    getRankList(currentDataType);
                }
                break;
            case 6:
                if (currentDataType != 5) {
                    currentDataType = 5;
                    getRankList(currentDataType);
                }
                break;
            case 7:
                if (currentDataType != 6) {
                    currentDataType = 6;
                    getRankList(currentDataType);
                }
                break;
            default:
                break;
        }
        Common.showLog(currentDataType);
    };

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_rank, container, false);
        initView(view);
        getFollowUserNewIllust();
        return view;
    }

    private void initView(View view) {
        mSharedPreferences = Common.getLocalDataSet(mContext);
        mRecyclerView = view.findViewById(R.id.pixiv_recy);
        GridLayoutManager gridLayoutManager = new GridLayoutManager(mContext, 2);
        gridLayoutManager.setSpanSizeLookup(new GridLayoutManager.SpanSizeLookup() {
            @Override
            public int getSpanSize(int position) {
                if (mPixivAdapter.getItemViewType(position) == 2) {
                    return gridLayoutManager.getSpanCount();
                } else {
                    return 1;
                }
            }
        });
        toolbar = view.findViewById(R.id.toolbar_pixiv);
        ((AppCompatActivity) Objects.requireNonNull(getActivity())).setSupportActionBar(toolbar);
        toolbar.setNavigationOnClickListener(view1 ->
                ((MainActivity) Objects.requireNonNull(getActivity()))
                        .getDrawer().openDrawer(Gravity.START, true));
        toolbar.setTitle("动态");
        mRecyclerView.setLayoutManager(gridLayoutManager);
        mRecyclerView.setHasFixedSize(true);
        mProgressBar = view.findViewById(R.id.try_login);
        bmb = view.findViewById(R.id.bmb);
        bmb.setUse3DTransformAnimation(true);
        bmb.setShowDuration(400);
        bmb.setHideDuration(400);
        bmb.setFrames(60);
        bmb.setNormalColor(getResources().getColor(R.color.colorAccent));
        for (int i = 0; i < bmb.getPiecePlaceEnum().pieceNumber(); i++) {
            TextInsideCircleButton.Builder builder = new TextInsideCircleButton.Builder()
                    .normalImageRes(R.drawable.ic_card_giftcard_black_24dp)
                    .imagePadding(new Rect(20, 20, 20, 60))
                    .normalText(arrayOfRankMode[i])
                    .textRect(new Rect(Util.dp2px(15), Util.dp2px(42), Util.dp2px(65), Util.dp2px(72)))
                    .textSize(16)
                    .listener(clickListener);
            bmb.addBuilder(builder);
        }
    }

    private void getFollowUserNewIllust() {
        mProgressBar.setVisibility(View.VISIBLE);
        Call<IllustfollowResponse> call = new RestClient()
                .getRetrofit_AppAPI()
                .create(AppApiPixivService.class)
                .getFollowIllusts(mSharedPreferences.getString("Authorization", ""), "public");
        call.enqueue(new Callback<IllustfollowResponse>() {
            @Override
            public void onResponse(@NonNull Call<IllustfollowResponse> call,
                                   @NonNull retrofit2.Response<IllustfollowResponse> response) {
                IllustfollowResponse illustfollowResponse = response.body();
                next_url = illustfollowResponse.getNext_url();
                initAdapter(illustfollowResponse.getIllusts());
                toolbar.setTitle("动态");
                mProgressBar.setVisibility(View.INVISIBLE);
            }

            @Override
            public void onFailure(@NonNull Call<IllustfollowResponse> call, @NonNull Throwable throwable) {

            }
        });
    }

    private void getRankList(int dataType) {
        mProgressBar.setVisibility(View.VISIBLE);
        Call<IllustRankingResponse> call = new RestClient()
                .getRetrofit_AppAPI()
                .create(AppApiPixivService.class)
                .getIllustRanking(mSharedPreferences.getString("Authorization", ""),
                        modelist[dataType], null);
        call.enqueue(new Callback<IllustRankingResponse>() {
            @Override
            public void onResponse(@NonNull Call<IllustRankingResponse> call,
                                   @NonNull retrofit2.Response<IllustRankingResponse> response) {
                Reference.sIllustRankingResponse = response.body();
                next_url = Reference.sIllustRankingResponse.getNext_url();
                initAdapter(Reference.sIllustRankingResponse.getIllusts());
                toolbar.setTitle(arrayOfRankMode[currentDataType + 1]);
                mProgressBar.setVisibility(View.INVISIBLE);
            }

            @Override
            public void onFailure(@NonNull Call<IllustRankingResponse> call, @NonNull Throwable throwable) {

            }
        });
    }

    private void getNextData() {
        if (next_url != null) {
            if (Reference.sIllustRankingResponse != null) {
                Reference.sIllustRankingResponse = null;
            }
            mProgressBar.setVisibility(View.VISIBLE);
            Call<RecommendResponse> call = new RestClient()
                    .getRetrofit_AppAPI()
                    .create(AppApiPixivService.class)
                    .getNext(mSharedPreferences.getString("Authorization", ""), next_url);
            call.enqueue(new Callback<RecommendResponse>() {
                @Override
                public void onResponse(@NonNull Call<RecommendResponse> call,
                                       @NonNull retrofit2.Response<RecommendResponse> response) {
                    Reference.sRankList = response.body();
                    next_url = Reference.sRankList.getNext_url();
                    initAdapter(Reference.sRankList.getIllusts());
                    mProgressBar.setVisibility(View.INVISIBLE);
                }

                @Override
                public void onFailure(@NonNull Call<RecommendResponse> call, @NonNull Throwable throwable) {

                }
            });
        } else {
            Snackbar.make(mProgressBar, "再怎么找也找不到了~", Snackbar.LENGTH_SHORT).show();
        }
    }

    private void initAdapter(List<IllustsBean> illustsBeans) {
        mPixivAdapter = new PixivAdapterGrid(illustsBeans, mContext);
        mPixivAdapter.setOnItemClickListener(new OnItemClickListener() {
            @Override
            public void onItemClick(View view, int position, int viewType) {
                if (position == -1) {
                    getNextData();
                } else if (viewType == 0) {
                    Reference.sIllustsBeans = illustsBeans;
                    Intent intent = new Intent(mContext, ViewPagerActivity.class);
                    intent.putExtra("which one is selected", position);
                    mContext.startActivity(intent);
                } else if (viewType == 1) {
                    if (!illustsBeans.get(position).isIs_bookmarked()) {
                        ((ImageView) view).setImageResource(R.drawable.ic_favorite_white_24dp);
                        view.startAnimation(Common.getAnimation());
                        Common.postStarIllust(position, illustsBeans,
                                mSharedPreferences.getString("Authorization", ""), mContext, "public");
                    } else {
                        ((ImageView) view).setImageResource(R.drawable.ic_favorite_border_black_24dp);
                        view.startAnimation(Common.getAnimation());
                        Common.postUnstarIllust(position, illustsBeans,
                                mSharedPreferences.getString("Authorization", ""), mContext);
                    }
                }
            }

            @Override
            public void onItemLongClick(View view, int position) {
                if (!illustsBeans.get(position).isIs_bookmarked()) {
                    ((ImageView) view).setImageResource(R.drawable.ic_favorite_white_24dp);
                    Common.postStarIllust(position, illustsBeans,
                            mSharedPreferences.getString("Authorization", ""), mContext, "private");
                }
            }
        });
        mRecyclerView.setAdapter(mPixivAdapter);
    }

    @Override
    public void onCreate(Bundle b) {
        super.onCreate(b);
        setHasOptionsMenu(true);
    }

    @Override
    public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
        super.onCreateOptionsMenu(menu, inflater);
        inflater.inflate(R.menu.main_pixiv, menu);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        Intent intent;
        switch (item.getItemId()) {
            case R.id.action_search:
                intent = new Intent(mContext, SearchActivity.class);
                mContext.startActivity(intent);
                return true;
            case R.id.action_settings:
                intent = new Intent(mContext, SettingsActivity.class);
                mContext.startActivity(intent);
                return true;
            default:
                return super.onOptionsItemSelected(item);
        }
    }

    @Override
    public void onResume() {
        super.onResume();
        if (mPixivAdapter != null) {
            mPixivAdapter.notifyDataSetChanged();
        }
    }
}

File: app/src/main/java/com/example/administrator/essim/fragments/FragmentRank.kt
package com.example.administrator.essim.fragments

import android.content.Intent
import android.graphics.Rect
import android.os.Bundle
import android.support.design.widget.Snackbar
import android.support.v7.app.AppCompatActivity
import android.support.v7.widget.GridLayoutManager
import android.view.*
import android.widget.ImageView
import com.example.administrator.essim.R
import com.example.administrator.essim.activities.MainActivity
import com.example.administrator.essim.activities.SearchActivity
import com.example.administrator.essim.activities.SettingsActivity
import com.example.administrator.essim.activities.ViewPagerActivity
import com.example.administrator.essim.adapters.PixivAdapterGrid
import com.example.administrator.essim.interf.OnItemClickListener
import com.example.administrator.essim.network.AppApiPixivService
import com.example.administrator.essim.network.RestClient
import com.example.administrator.essim.response.*
import com.example.administrator.essim.utils.Common
import com.nightonke.boommenu.BoomButtons.TextInsideCircleButton
import com.nightonke.boommenu.Util
import kotlinx.android.synthetic.main.fragment_rank.*
import retrofit2.Call
import retrofit2.Callback


/**
 * Created by Administrator on 2018/1/15 0015.
 */

class FragmentRank : BaseFragment() {

    private val arrayOfRankMode = arrayOf("动态", "日榜", "周榜", "月榜", "新人", "原创", "男性向", "女性向")
    private val modeList = arrayOf("day", "week", "month", "week_rookie", "week_original", "day_male", "day_female")
    private var currentDataType = -1
    private var nextDataUrl: String? = null
    private var mPixivAdapter: PixivAdapterGrid? = null
    private val clickListener = { index: Int ->
        when (index) {
            0 -> if (currentDataType != -1) {
                currentDataType = -1
                getFollowUserNewIllust()
            }
            1 -> if (currentDataType != 0) {
                currentDataType = 0
                getRankList(currentDataType)
            }
            2 -> if (currentDataType != 1) {
                currentDataType = 1
                getRankList(currentDataType)
            }
            3 -> if (currentDataType != 2) {
                currentDataType = 2
                getRankList(currentDataType)
            }
            4 -> if (currentDataType != 3) {
                currentDataType = 3
                getRankList(currentDataType)
            }
            5 -> if (currentDataType != 4) {
                currentDataType = 4
                getRankList(currentDataType)
            }
            6 -> if (currentDataType != 5) {
                currentDataType = 5
                getRankList(currentDataType)
            }
            7 -> if (currentDataType != 6) {
                currentDataType = 6
                getRankList(currentDataType)
            }
            else -> {
            }
        }
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        return inflater.inflate(R.layout.fragment_rank, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        initView()
        getFollowUserNewIllust()
    }

    private fun initView() {
        val gridLayoutManager = GridLayoutManager(mContext, 2)
        gridLayoutManager.spanSizeLookup = object : GridLayoutManager.SpanSizeLookup() {
            override fun getSpanSize(position: Int): Int {
                return when {
                    mPixivAdapter!!.getItemViewType(position) == 2 -> gridLayoutManager.spanCount
                    else -> 1
                }
            }
        }
        (activity as AppCompatActivity).setSupportActionBar(mToolbar)
        mToolbar.setNavigationOnClickListener { (activity as MainActivity).drawer.openDrawer(Gravity.START, true) }
        mToolbar.title = "动态"
        mRecyclerView.layoutManager = gridLayoutManager
        mRecyclerView.setHasFixedSize(true)
        mFab.isUse3DTransformAnimation = true
        mFab.showDuration = 400
        mFab.hideDuration = 400
        mFab.frames = 60
        mFab.normalColor = resources.getColor(R.color.colorAccent)
        for (i in 0 until mFab.piecePlaceEnum.pieceNumber()) {
            val builder = TextInsideCircleButton.Builder()
                    .normalImageRes(R.drawable.ic_card_giftcard_black_24dp)
                    .imagePadding(Rect(20, 20, 20, 60))
                    .normalText(arrayOfRankMode[i])
                    .textRect(Rect(Util.dp2px(15f), Util.dp2px(42f), Util.dp2px(65f), Util.dp2px(72f)))
                    .textSize(16)
                    .listener(clickListener)
            mFab.addBuilder(builder)
        }
    }

    private fun getFollowUserNewIllust() {
        mProgressbar.visibility = View.VISIBLE
        val call = RestClient()
                .retrofit_AppAPI
                .create(AppApiPixivService::class.java)
                .getFollowIllusts(Common.getLocalDataSet().getString("Authorization", "")!!, "public")
        call.enqueue(object : Callback<IllustfollowResponse> {
            override fun onResponse(call: Call<IllustfollowResponse>,
                                    response: retrofit2.Response<IllustfollowResponse>) {
                val illustfollowResponse = response.body()
                nextDataUrl = illustfollowResponse!!.next_url
                initAdapter(illustfollowResponse.illusts)
                mToolbar.title = "动态"
                mProgressbar.visibility = View.INVISIBLE
            }

            override fun onFailure(call: Call<IllustfollowResponse>, throwable: Throwable) {}
        })
    }

    private fun getRankList(dataType: Int) {
        mProgressbar.visibility = View.VISIBLE
        val call = RestClient()
                .retrofit_AppAPI
                .create(AppApiPixivService::class.java)
                .getIllustRanking(Common.getLocalDataSet().getString("Authorization", "")!!,
                        modeList[dataType], null)
        call.enqueue(object : Callback<IllustRankingResponse> {
            override fun onResponse(call: Call<IllustRankingResponse>,
                                    response: retrofit2.Response<IllustRankingResponse>) {
                nextDataUrl = response.body()!!.next_url
                initAdapter(response.body()!!.illusts)
                mToolbar.title = arrayOfRankMode[currentDataType + 1]
                mProgressbar.visibility = View.INVISIBLE
            }

            override fun onFailure(call: Call<IllustRankingResponse>, throwable: Throwable) {}
        })
    }

    private fun getNextData() {
        if (nextDataUrl != null) {
            mProgressbar.visibility = View.VISIBLE
            val call = RestClient()
                    .retrofit_AppAPI
                    .create(AppApiPixivService::class.java)
                    .getNext(Common.getLocalDataSet().getString("Authorization", "")!!, nextDataUrl!!)
            call.enqueue(object : Callback<RecommendResponse> {
                override fun onResponse(call: Call<RecommendResponse>,
                                        response: retrofit2.Response<RecommendResponse>) {
                    nextDataUrl = response.body()!!.next_url
                    initAdapter(response.body()!!.illusts)
                    mProgressbar.visibility = View.INVISIBLE
                }

                override fun onFailure(call: Call<RecommendResponse>, throwable: Throwable) {}
            })
        } else {
            Snackbar.make(mProgressbar, "再怎么找也找不到了~", Snackbar.LENGTH_SHORT).show()
        }
    }

    private fun initAdapter(illustsBeans: List<IllustsBean>) {
        mPixivAdapter = PixivAdapterGrid(illustsBeans, mContext)
        mPixivAdapter!!.setOnItemClickListener(object : OnItemClickListener {
            override fun onItemClick(view: View, position: Int, viewType: Int) {
                when {
                    position == -1 -> getNextData()
                    viewType == 0 -> {
                        Reference.sIllustsBeans = illustsBeans
                        val intent = Intent(mContext, ViewPagerActivity::class.java)
                        intent.putExtra("which one is selected", position)
                        mContext.startActivity(intent)
                    }
                    viewType == 1 -> if (!illustsBeans[position].isIs_bookmarked) {
                        (view as ImageView).setImageResource(R.drawable.ic_favorite_white_24dp)
                        view.startAnimation(Common.getAnimation())
                        Common.postStarIllust(position, illustsBeans,
                                Common.getLocalDataSet().getString("Authorization", ""), mContext, "public")
                    } else {
                        (view as ImageView).setImageResource(R.drawable.ic_favorite_border_black_24dp)
                        view.startAnimation(Common.getAnimation())
                        Common.postUnstarIllust(position, illustsBeans,
                                Common.getLocalDataSet().getString("Authorization", ""), mContext)
                    }
                }
            }

            override fun onItemLongClick(view: View, position: Int) {
                if (!illustsBeans[position].isIs_bookmarked) {
                    (view as ImageView).setImageResource(R.drawable.ic_favorite_white_24dp)
                    Common.postStarIllust(position, illustsBeans,
                            Common.getLocalDataSet().getString("Authorization", ""), mContext, "private")
                }
            }
        })
        mRecyclerView.adapter = mPixivAdapter
    }

    override fun onCreate(b: Bundle?) {
        super.onCreate(b)
        setHasOptionsMenu(true)
    }

    override fun onCreateOptionsMenu(menu: Menu?, inflater: MenuInflater?) {
        super.onCreateOptionsMenu(menu, inflater)
        inflater!!.inflate(R.menu.main_pixiv, menu)
    }

    override fun onOptionsItemSelected(item: MenuItem?): Boolean {
        val intent: Intent
        return when (item!!.itemId) {
            R.id.action_search -> {
                intent = Intent(mContext, SearchActivity::class.java)
                mContext.startActivity(intent)
                true
            }
            R.id.action_settings -> {
                intent = Intent(mContext, SettingsActivity::class.java)
                mContext.startActivity(intent)
                true
            }
            else -> super.onOptionsItemSelected(item)
        }
    }

    override fun onResume() {
        super.onResume()
        if (mPixivAdapter != null) {
            mPixivAdapter!!.notifyDataSetChanged()
        }
    }
}